src/lib/mesg_client.cpp

```diff
@@ -153,14 +156,15 @@ std::error_condition mesg_factory::reinit_client(const std::string& client,
     return std::error_condition();
 }
 
-std::error_condition mesg_factory::data_service_request(std::string const& request_name, io_blob_list_t const& cli_buf,
-                                                        data_service_response_handler_t const& response_cb) {
+AsyncResult< sisl::io_blob > mesg_factory::data_service_request(std::string const& request_name,
+                                                                io_blob_list_t const& cli_buf) {
     std::shared_lock< client_factory_lock_type > rl(_client_lock);
+    auto calls = std::list< AsyncResult< sisl::io_blob > >();
     for (auto& nuraft_client : _clients) {
         auto g_client = std::dynamic_pointer_cast< nuraft_mesg::group_client >(nuraft_client.second);
-        g_client->data_service_request(get_generic_method_name(request_name, _group_name), cli_buf, response_cb);
+        calls.push_back(g_client->data_service_request(get_generic_method_name(request_name, _group_name), cli_buf));
```

Comment:
2 questions:
1. Should we really need a std::list here, can we have a thread_local std::vector or sisl::VectorPool to avoid STL internal memory allocation
2. I haven't fully grasped the call trace, but if we are doing collectAny and let caller do the assimilation, can we not put the array of promises to the caller and let them do wait on all or wait for quorum etc..

Reply:
I need to go through the data_svc stuff still a little. It was added at the beginning of our NuObject work which I only glanced at. Will take a look :+1:

Reply:
Originally, `data_service_request` method was intended to call the method to all the clients in the raft group. We do not need to wait for quorum here, each call is independent because the quorum is taken care by the raft path. Still, this method is not sufficient as we need a function `data_request_to_one` when the follower needs to request data .

Reply:
@szmyd I am not fully clear on the folly future terminology, do we need to collect the results if they are independent to each other?

Reply:
We are using `collectAny` which will unblock as soon as we get back a single result.

Reply:
I am confused why this method (mesg_factory::data_service_request) is returning AsyncResult< sisl::io_blob >. This method just iterates over all clients and makes an async call to each one of the clients. These calls are independent

Reply:
previously, each of these async calls to the clients had their own callback called and this method returned std::error_condition